save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/LnkUserProjectController.kt

```diff
@@ -7,100 +7,156 @@
-     * @throws NoSuchElementException
-     * @throws ResponseStatusException
-     */
     @GetMapping("/{organizationName}/{projectName}/users/not-from")
-    @Suppress("UnsafeCallOnNullableType")
+    @RequiresAuthorizationSourceHeader
+    @PreAuthorize("isAuthenticated()")
+    @Operation(
+        method = "GET",
+        summary = "Get users not from project.",
+        description = "Get list of users that are not connected with given project.",
+    )
+    @Parameters(
+        Parameter(name = "organizationName", `in` = ParameterIn.PATH, description = "name of an organization", required = true),
+        Parameter(name = "projectName", `in` = ParameterIn.PATH, description = "name of a project", required = true),
+    )
+    @ApiResponse(responseCode = "200", description = "Successfully fetched users not from project.")
+    @ApiResponse(responseCode = "404", description = "Project with such name was not found or considered to be private.")
+    @Suppress("TOO_LONG_FUNCTION")
     fun getAllUsersNotFromProjectWithNamesStartingWith(
         @PathVariable organizationName: String,
         @PathVariable projectName: String,
         @RequestParam prefix: String,
         authentication: Authentication,
-    ): List<UserInfo> {
-        if (prefix.isEmpty()) {
-            return emptyList()
+    ): Mono<List<UserInfo>> = Mono.zip(
+        Mono.just(organizationName),
+        Mono.just(projectName),
+    )
+        .filter {
+            prefix.isNotEmpty()
         }
-        val project = projectService.findByNameAndOrganizationName(projectName, organizationName)
-            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
-        if (!projectPermissionEvaluator.hasPermission(authentication, project, Permission.READ)) {
-            throw ResponseStatusException(HttpStatus.NOT_FOUND)
+        .flatMap { (organizationName, projectName) ->
+            projectService.findByNameAndOrganizationName(projectName, organizationName).toMono()
         }
-        val projectUserIds = lnkUserProjectService.getAllUsersByProject(project).map { it.id!! }.toSet()
-        // first we need to get users with exact match by name
-        val exactMatchUsers = lnkUserProjectService.getNonProjectUsersByName(prefix, projectUserIds)
-        // and then we need to get those whose names start with `prefix`
-        val prefixUsers = lnkUserProjectService.getNonProjectUsersByNamePrefix(
-            prefix,
-            projectUserIds + exactMatchUsers.map { it.id!! },
-            PAGE_SIZE - exactMatchUsers.size,
-        )
-        return (exactMatchUsers + prefixUsers).map { it.toUserInfo() }
-    }
+        .switchIfEmptyToNotFound {
+            "No project with name $projectName was found in organization $organizationName."
+        }
+        .filter { project ->
+            projectPermissionEvaluator.hasPermission(authentication, project, Permission.READ)
+        }
+        .switchIfEmptyToNotFound {
```

Comment:
maybe forbidden instead of not found?

Reply:
We decided that private projects should be invisible just as they don't exist